numba/cuda/cudadrv/devices.py

```diff
@@ -1,134 +1,228 @@
+class _Runtime(object):
+    """Emulate the CUDA runtime context management.
+
+    It owns all Devices and Contexts.
+    Keeps at most one Context per Device
     """
 
-    def __init__(self, gpu):
-        self._gpu = gpu
-        self._context = None
+    def __init__(self):
+        self.gpus = _DeviceList()
 
-    def __getattr__(self, key):
-        """Redirect to self._gpu
-        """
-        if key.startswith('_'):
-            raise AttributeError(key)
-        return getattr(self._gpu, key)
+        # A thread local stack
+        self.context_stack = servicelib.TLStack()
+        # Remembers all context
+        # key: context.handle.value
+        # value: Context
+        self._contexts = {}
 
-    def __repr__(self):
-        return repr(self._gpu)
+        # Device's context map
+        # key: Device.id
+        # value: Context
+        self._devctx = {}
 
-    def associate_context(self):
-        """Associate the context of this GPU to the running thread
+        # Remember the main thread
+        # Only the main thread can *actually* destroy
+        self._mainthread = threading.current_thread()
+
+        # Avoid mutation of runtime state in multithreaded programs
+        self._lock = threading.Lock()
+
+    @property
+    def current_context(self):
+        """Return the active gpu context
         """
-        # No context was created for this GPU
-        if self._context is None:
-            self._context = self._gpu.create_context()
+        top = self.context_stack.top
+        # integrity check
+        assert driver.get_context().value == top.handle.value, (
+            "An unmanaged CUDA context is active."
+        )
+        return top
+
+    def _create_context(self, gpu):
+        """Create a new context for the given gpu
+        """
+        # Use lock to prevent internal state mutation
+        with self._lock:
+            ctx = gpu.create_context()
+            self._contexts[ctx.handle.value] = ctx
+            self._devctx[gpu.id] = ctx
 
-        # Current context is not associated with the thread
-        if self._gpu.get_context() != self._context:
-            self._context.push()
+            assert len(self._contexts) <= len(gpus), _MSG_TOO_MANY_CTX_PER_DEV
+            assert len(self._devctx) <= len(gpus), _MSG_TOO_MANY_CTX_PER_DEV
+            return ctx
 
-        return self._context
+    def _get_or_create_context(self, gpu):
+        """Try to use a already created context for the given gpu.  If none
+        existed, create a new context.
 
-    def __enter__(self):
-        self.associate_context()
-        _gpustack.push(self)
+        Returns the context
+        """
+        # Get a context from the device
+        ctx = self._devctx.get(gpu.id)
+        if ctx is None:
+            ctx = self._create_context(gpu)
+        else:
+            # Push the context to the stack
+            ctx.push()
+
+        return ctx
+
+    def push_context(self, gpu):
```

Comment:
It looks like `push_context` will always push a context for the given GPU on the stack regardless of whether it is the current context - is that supposed to happen, or am I misunderstanding something?